draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+compression or decompression, encryption or decryption etc.
+In some cases where no asynchronous code for the I/O operation is available
+CPU bound tasks are also an appropriate choice as they will not block
+the event loop for the duration of the task.
+
+Performing operations which release the :term:`GIL` is recommended to avoid
+throttling the concurrency of the worker.
+
+CPU bound tasks are specifically marked as such using Python's
+`def` notation for defining functions. They will run in a Python thread.
+
+Using threads instead of forking the main process has its upsides:
+
+- It simplifies the Worker's architecture and makes it less brittle.
+
+  Processes require :term:`IPC` to communicate with each other.
+  This complicates implementation since multiple methods are required to support
+  :term:`IPC` reliably across all operating systems Celery supports.
+  Threads on the other hand require less complicated means of communication.
+
+  In `trio`_, we simply use a memory channel which is a coroutine and thread safe
+  way to send and receive values.
+
+- PyPy's JIT warms up faster.
+
+  When using PyPy, using threads means that we get to keep our previous JIT traces
+  and therefore JIT warmup will occur faster.
+
+  If we'd use processes, each process has to warm up its own JIT which results
+  in tasks being executed slower for a longer period of time.
+
+Using threads for CPU bound tasks unfortunately has some downsides as well:
+
+- Pure Python CPU bound workloads cannot be executed in parallel.
+
+  In both CPython and PyPy the :term:`GIL` prevents executing two Python bytecodes
+  in parallel by design.
+
+  This results in slower execution of Python code when using threads.
+
+- The :term:`GIL`'s implementation in CPython 3.x has a defect in design.
+
+  According to a `bug report <https://bugs.python.org/issue7946>`_ the new GIL
+  in Python 3 CPU bound threads may starve I/O threads (in our case the main thread).
+
+  .. note ::
+
+    This is not an issue with PyPy's implementation of the `GIL`_
+    `according to Armin Rigo <https://bugs.python.org/msg346495>`_, PyPy's creator.
+
+  - Tasks are no longer isolated.
+
+    Since we're mixing workloads to maximize our throughput a task which crashes
+    the worker or leaks memory can crash the entire worker.
+
+To mitigate these issues CPU Bound Tasks may be globally rate limited to allow
+the main thread to complete executing :ref:`draft/celery-5-high-level-architecture:I/O Bound Tasks`.
+
+Boxed Tasks
+~~~~~~~~~~~
+
+To minimize the disadvantages of using threads in Python and workaround the
+shortcomings of the :term:`GIL`, Celery also provides a new type of tasks called
+Boxed Tasks.
+
+Boxed Tasks are processes which execute tasks in an isolated manner.
+
+The processes' lifecycle is managed by the :ref:`draft/celery-5-high-level-architecture:Controller`.
+
+Since Boxed Tasks are run separately from Celery itself, the program the process
```

Comment:
Not sure I comprehend this part. Will Celery start a new process by running some executable and pass information to it through some communication channel?

Reply:
Yes.
The Controller is responsible for starting Boxed Workers for Boxed Tasks.

Reply:
Does that resemble [Airflow](https://airflow.apache.org/) in some part?